File: src/19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
'''
不用双指针，而是用数字记录长度
'''
#time defeat-82.76%
#space defeat-8.92%
class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        #极端情况
        #边际情况，0-1-2个节点
        #空链表
        if not head:
            return head
        #1个节点
        elif not head.next:
            return None
        #两个节点
        elif not head.next.next:
            #根据n返回节点
            if n == 1:
                head.next = None
                return head
            else:
                return head.next
        else:
            #大于三个节点的时候
            #先获取链表长度
            listLen = 0
            nowNode = head  #指向头部的另一个指针
            while nowNode != None:
                listLen += 1
                nowNode = nowNode.next
            #应对只有三个的情况
            if listLen == n:
                return head.next
            #然后第二遍扫描链表
            nowNode = head
            nowLen = listLen - n - 1
            while nowLen > 0:
                nowNode = nowNode.next
                nowLen -= 1
            #停下来时候，nowNode指向被删除节点的前一个
            nowNode.next = nowNode.next.next
            return head
                
        

File: src/19.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        //干翻阿里巴巴
        //两个指针，一个落后另一个n个，当另一个到末尾时，就可以删除了
        //一次扫描结束
        //处理边界情况
        //十分钟，一次提交，卧槽，新冠就在我身边
        //时间击败-100%，空间击败-76.83%
        ListNode * fast = head;
        while(n-- && fast != nullptr){
            fast = fast->next;
        }
        //如果fast为空，则n肯定是第一个节点，返回
        if(fast == nullptr){
            ListNode * result = head->next;
            delete head;
            return result;
        }
        ListNode * prev = head;
        while(fast->next != nullptr){
            fast = fast->next;
            prev = prev->next;
        }
        //到达末尾
        ListNode * target = prev->next;
        //删除
        prev->next = target->next;
        delete target;
        return head;
    }
};
